File: user-auth-service/src/main/java/com/damian/marketgrid/user/messaging/KafkaTopic.java
package com.damian.marketgrid.user.messaging;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Getter
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public enum KafkaTopic {
    USER_CHANGED("user-changed");

    private final String topic;
}

File: ingest-service/src/main/kotlin/com/damian/marketgrid/messaging/KafkaTopic.kt
package com.damian.marketgrid.messaging

import com.damian.marketgrid.messaging.event.CurrencyRatesIngestedEvent
import com.damian.marketgrid.messaging.event.MarketResourceIngestedEvent

enum class KafkaTopic(
    val topic: String,
    val eventClass: Class<*>
) {
    MARKET_RESOURCE_INGESTED("market-resource-ingested", MarketResourceIngestedEvent::class.java),
    CURRENCY_RATES_INGESTED("currency-rates-ingested", CurrencyRatesIngestedEvent::class.java);

    companion object {
        fun of(obj: Any): KafkaTopic =
            entries.find { it.eventClass.isInstance(obj) }
                ?: throw IllegalArgumentException("No Kafka topic exists for object of type: ${obj::class.java.name}")
    }
}
